worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,1062 @@
+	/// and their last blocks in memory for better performance
+	pub fn new(path: PathBuf) -> Result<SidechainStorage<SignedBlock>> {
+		// load db
+		let db = SidechainDB::open_default(path)?;
+		let mut storage = SidechainStorage { db, shards: vec![], last_blocks: HashMap::new() };
+		storage.shards = storage.load_shards_from_db()?;
+		// get last block of each shard
+		for shard in storage.shards.iter() {
+			if let Some(last_block) = storage.load_last_block_from_db(shard)? {
+				storage.last_blocks.insert(*shard, last_block);
+			} else {
+				// an empty shard sidechain storage should not exist. Consider deleting this shard from the shards list.
+				error!("Sidechain storage of shard {:?} is empty", shard);
+			}
+		}
+		Ok(storage)
+	}
+
+	/// gets all shards of currently loaded sidechain db
+	pub fn shards(&self) -> &Vec<ShardIdentifierFor<SignedBlock>> {
+		&self.shards
+	}
+
+	/// gets the last block of the current sidechain DB and the given shard
+	pub fn last_block_of_shard(
+		&self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+	) -> Option<&LastSidechainBlock> {
+		self.last_blocks.get(shard)
+	}
+
+	/// gets the block hash of the sidechain block of the given shard and block number, if there is such a block
+	pub fn get_block_hash(
+		&self,
+		shard: &ShardIdentifierFor<SignedBlock>,
+		block_number: BlockNumber,
+	) -> Result<Option<BlockHash>> {
+		self.db.get((*shard, block_number))
+	}
+
+	/// gets the block of the given blockhash, if there is such a block
+	#[allow(unused)]
+	pub fn get_block(&self, block_hash: &BlockHash) -> Result<Option<SignedBlock>> {
+		self.db.get(block_hash)
+	}
+
+	/// update sidechain storage
+	pub fn store_blocks(&mut self, blocks_to_store: Vec<SignedBlock>) -> Result<()> {
+		let mut batch = WriteBatch::default();
+		let mut new_shard = false;
+		for signed_block in blocks_to_store.into_iter() {
+			// check if current block is the next in line
+			let current_shard = &signed_block.block().shard_id();
+			let current_block_nr = signed_block.block().block_number();
+			if self.shards.contains(current_shard) {
+				if let Some(last_block) = self.last_block_of_shard(current_shard) {
+					if last_block.number != current_block_nr - 1 {
+						error!("[Sidechain DB] Sidechainblock (nr: {:?}) is not a succession of the previous block (nr: {:?}) in shard: {:?}",
+						current_block_nr, last_block.number, *current_shard);
+						continue
+					}
+				} else {
+					error!(
+						"[Sidechain DB] Shard {:?} does not have a last block. Skipping block (nr: {:?}) inclusion",
+						*current_shard, current_block_nr
+					);
+					continue
+				}
+			} else {
+				self.shards.push(*current_shard);
+				new_shard = true;
+			}
+			// add block to DB batch
+			self.add_last_block(&mut batch, &signed_block);
+		}
+		// update stored_shards_key -> vec<shard> only if a new shard was included
+		if new_shard {
+			SidechainDB::add_to_batch(&mut batch, STORED_SHARDS_KEY, self.shards().clone());
```

Comment:
This is really really a tiny detail, but this should not be a static method of the SidechainDB but rather an extension trait for the batch, so that you can do: `bach.add(Key, shard)`.

Reply:
So you want me to write an extra Struct to wrap the `WriteBatch`?

Reply:
Now you would not need a struct to wrap it. If you define a trait you can directly implement it for `WriteBatch`.

But this was more a hint of what is possible. I am also fine if we leave it as is.

Reply:
Ah, I see. That makes sense. Thanks for the input